### PR TITLE
Adaptive patch2self

### PR DESCRIPTION
**Pull request by captainnova**
Hello, this will probably be most interesting to @ShreyasFadnavis .

Shreyas, I think I figured out why p2s is blurring in diffusion space, and mostly fixed it. The problem is that by training the regressor to predict an entire volume it is assuming that all of the voxels have the same "diffusion response function", i.e. it is like saying they have the same diffusion tensor. The result looks better than that, though, because the regressor is free to give more weight to similar volumes, so even though it is using a one-size-fits-all interpolation kernel, it does not have to interpolate from very far away.

My solution is to train different regressors for different types of voxels, and then for any given voxel use a weighted sum of the regressors. The "types of voxels" comes from either Independent Component Analysis or Principal Component Analysis (ICA is slightly better), and the voxel's position in component space also sets the weights for the regressors. Thus each voxel gets a regressor that is appropriate to it, preserving signal without hurting denoising by going to the extreme of giving each voxel a regressor that is exactly tuned to it.

Since the equivalent of a "patch" is now a collection of voxels with similar intensities for each volume, regardless of their spatial location, I took out patch_radius. I now think p2s really is better working only in diffusion space, and the improvement I was seeing before with larger patches was just from the regressor getting more variables to work with. Because the interface changed, I ended up making adaptive_patch2self its own command instead of trying to insert this into p2s.

This will probably make more sense with pictures, but I'll submit the pull request for review now and add images later instead of trying to do everything at once.

Best wishes,

Rob

### CODE REVIEW

**Comment by captainnova** [comment]
I forgot to mention that another reason to drop patch_radius is that I added masking, which gives a big speed boost. So altogether with a few x speedup from masking, and a couple of dozen more regressors, adaptive_patch2self is only several x slower than patch2self.

**Comment by ShreyasFadnavis** [comment]
@captainnova:
Hi Rob, this is very good work and thanks a lot for the PR! I am looking at your code and looks quite good to me. I like the idea name Patch2Self Adaptive too 👍🏽 

I only briefly looked at the code and realized that it could be integrated withing Patch2Self as an option. We can of course keep the workflow separate -- which looks good to me. For the `patch_radius=0` I was going to downgrade that in fashion to what you have too! 

I also have some new work on speeding up Patch2Self which would work with your modification too :) I will try the code out ASAP (especially with the data that you have). Can you also add a small example of how I can run the code and reproduce your results... So that we are on the same page? It will help with reviewing the PR too 👍🏽 

As a side question: do you get the same result with and without masking? If yes -- That is quite neat!

**Comment by captainnova** [comment]
@ShreyasFadnavis :

Hi Shreyas, thanks for the kind words.

> I only briefly looked at the code and realized that it could be easily integrated withing Patch2Self as an option.

That would make sense to me too, and was my original intent, but when the urge to get rid of patch_radius grew too strong I ended up splitting it into a new command, because I didn't want to "break" p2s. As the original author of p2s, though, you have more freedom in choosing what changes to introduce.

> The one that is in SKlearn is slowing things down I assume.

I haven't compared different implementations of the regressions or decompositions for speed. Using DGESVD for PCA and sklearn for ICA is a little weird, since sklearn also comes with PCA, and has a simpler/more consistent interface. But I started with PCA, and copied the example in denoise_lpca. When I added ICA I wondered about switching PCA to sklearn, but for all I know dipy has a reason to use DGESVD instead.

> I also have some new work on speeding up Patch2Self which would work with your modification too :) I will try the code out ASAP (especially with the data that you have). Can you also add a small example of how I can run the code and reproduce your results... So that we are on the same page? It will help with reviewing the PR too 👍🏽

Running it is really easy, and very similar to p2s. I added a dipy_denoise_adaptive_patch2self CLI (which I'm not sure I've used, to be honest), but here is the python gist:
```python
import dipy.denoise.adaptive_patch2self as ap2s
from dipy.io.image import load_nifti, save_nifti
import numpy as np

data, affine = load_nifti('raw.nii')
bvals = np.loadtxt('raw.bval')

# I use https://github.com/captainnova/dmri_segmenter to make these
mask, _ = load_nifti('raw_TIV.nii')

dnarr = ap2s.adaptive_patch2self, data, bvals, mask=mask, out_dtype=np.float32, verbose=True)
save_nifti('dnlinica.nii', dnarr, affine)
```

Properly evaluating the results takes more code, though. It is especially important to look at how the residuals change w.r.t. volume, not just the spatial pattern. Unfortunately the unwanted correlations do not show up very well in either movies of the residuals or time series of individual voxels. They do show up very well, and can be quantified, in maps of Pearson's r vs. signal, but that takes some code. What is the recommended way to share code that is used to investigate how a feature works, but is not needed by the feature itself?

> As a side question: do you get the same result with and without masking? If yes -- That is quite neat!

Almost - I think the result is slightly better with masking. What happens without masking is the first component is "is it air?", which is not very interesting for denoising. With masking that component goes away, and the others get promoted by one. But with 8-12 component axes being used in total this does not make a huge difference. It would probably make a bigger difference for original p2s. Without masking it is mostly learning the difference between air and not air, but with masking it would effectively be one component ahead of where it is now. You could simulate this by using ap2s(with and without masking, n_comps=0).

**Review by skoudoro — COMMENTED** [review]
Hi @captainnova,

Thank you for this new feature! 

I would like to know what is the next step concerning this PR @ShreyasFadnavis and @captainnova.

+1 for integrating within Patch2Self as an option

Could you precise if this PR is still in work in progress or not? 

If not, Could you add a short tutorial? It would make it easier to test. 

It would be good if you could improve the consistency of your docstring in general. Sometimes, it is really good, and sometimes inexistent. See below for a really quick review (just code style) until I know more about the status.

Thank you!

**Comment by skoudoro** [comment]
Also, Do not forget to rebase your PR since there is currently some conflict with the codebase.

Thank you !

**Comment by captainnova** [comment]
Hi @skoudoro,

> I would like to know what is the next step concerning this PR @ShreyasFadnavis and @captainnova.
> 
> +1 for integrating within Patch2Self as an option

That's exactly what @ShreyasFadnavis is working on. My PR was mostly a feature addition, but because I also took a couple of things out, I implemented it as a separate command. Shreyas agreed that patch2self no longer needs the removed items either, so adaptive_patch2self and patch2self are being merged.

I agree with your code style comments, but think it is best to wait for Shreyas to complete his merge + incorporate your comments + rebase, and then see what remains to be done.

File: dipy/denoise/adaptive_patch2self.py [hunk 1]
```diff
@@ -0,0 +1,699 @@
...
+
+if not has_sklearn:
+    w = "Scikit-Learn is required to denoise the data via (Adaptive)Patch2Self."
+    warn(w)
+
+
+def site_weight_beam(U, pca_ind, sign, sigma, growth_func):
+    """Return weights that increase going along the sign direction of
+    U[:, pca_ind], and fall off Gaussianly going away from that axis.
+    """
```
[the remark this answers is not in the document]
Comment (captainnova) [reply dipy/denoise/adaptive_patch2self.py:29]: agreed

File: dipy/denoise/adaptive_patch2self.py [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+
+def site_weight_beam_linear(U, pca_ind, sign, sigma):
+    return site_weight_beam(U, pca_ind, sign, sigma, lambda x: x)
+
+
+def site_weight_beam_arctan(U, pca_ind, sign, sigma):
+    return site_weight_beam(U, pca_ind, sign, sigma, np.arctan)
+
+
+def getRandomState(seed):
```
[the remark this answers is not in the document]
Comment (captainnova) [reply dipy/denoise/adaptive_patch2self.py:49]: I have no objection to underscored_function_names. @ShreyasFadnavis ?

File: dipy/denoise/adaptive_patch2self.py [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+    U : (nsamples, nfeatures) array
+        The principal components in sample space as orthonormal column vectors,
+        matching the order of S.
+    S : 1D array with len min(nsamples, nfeatures)
+        The "eigenvalues", from largest to smallest
+    Vt : (nfeatures, nfeatures) array
+        The principal components in feature space as orthonormal row vectors,
+        matching the order of S.
+    """
+    U, S, Vt = svd(X, *svd_args)[:3]
```
[the remark this answers is not in the document]
Comment (captainnova) [reply dipy/denoise/adaptive_patch2self.py:85]: I agree, but I copied this from localpca.py thinking it was dipy's preferred way of doing PCA/SVD. If there's no performance benefit to doing it that way, using sklearn.decomposition would provide a better interface.

File: dipy/denoise/adaptive_patch2self.py [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+
+
+def doICA(X, n_comps=None, seed=None, whiten=True):
+    """Decompose X into independent components with FastICA.
+
+    Parameters
+    ----------
+    X : (nsamples, nfeatures) array
+        The array to decompose.
+    n_comps : int or None
```
[the remark this answers is not in the document]
Comment (captainnova) [reply dipy/denoise/adaptive_patch2self.py:110]: OK

File: dipy/denoise/adaptive_patch2self.py [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+        The "eigenvalues", from largest to smallest
+    Vt : (nfeatures, nfeatures) array
+        The principal components in feature space as orthonormal row vectors,
+        matching the order of S.
+    """
+    ica = decomposition.FastICA(n_comps, random_state=getRandomState(seed))
+    return ica.fit(X).transform(X)
+
+
+class AdaptivePatch2Self(object):
```
[the remark this answers is not in the document]
Comment (captainnova) [reply dipy/denoise/adaptive_patch2self.py:132]: OK, I'm still necessarily in the habit of 2/3 compatibility. That's a nice link, gathering a lot of info in one place.

File: dipy/denoise/adaptive_patch2self.py [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+    weighted according to the argument (X)'s distance to each neighborhood.
+
+    In other words it is a version of Patch2Self where instead of just one set
+    of coefficients per volume being denoised, there is a collection of them,
+    and the regressor smoothly varies the weights of each to adapt to the
+    voxel's type (or mix of them), without allowing the regressor to exactly
+    fit the voxel, which would prevent denoising.
+    """
+    def __init__(self, data, n_comps=None, dtype=np.float32, model='ols',
+                 mod_kwargs={}, site_weight_func=site_weight_beam_linear,
```
[the remark this answers is not in the document]
Comment (captainnova) [reply dipy/denoise/adaptive_patch2self.py:146]: Good to know!

File: dipy/denoise/adaptive_patch2self.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (skoudoro) [reply dipy/denoise/adaptive_patch2self.py:85]: So, good catch! 😄  we need to change this in `localpca.py` also

**Comment by skoudoro** [comment]
ok, great! Thank you for the information! So let's wait for Shreyas, I will tag this PR as WIP to avoid any review before this update.

**Comment by ShreyasFadnavis** [comment]
> Hi @skoudoro,
> 
> > I would like to know what is the next step concerning this PR @ShreyasFadnavis and @captainnova.
> > +1 for integrating within Patch2Self as an option
> 
> That's exactly what @ShreyasFadnavis is working on. My PR was mostly a feature addition, but because I also took a couple of things out, I implemented it as a separate command. Shreyas agreed that patch2self no longer needs the removed items either, so adaptive_patch2self and patch2self are being merged.
> 
> I agree with your code style comments, but think it is best to wait for Shreyas to complete his merge + incorporate your comments + rebase, and then see what remains to be done.

+1 to what Rob said! I am testing the code and also working on integrating it with the current implementation. 
@skoudoro thank you for the review 💯 I will make my changes on top of this branch as soon as I have it working 👍🏽

**Comment by skoudoro** [comment]
Hi @captainnova, Hi @ShreyasFadnavis,

We plan to talk about your PR during the DIPY online meeting tomorrow (October 7th, 2021 | 1pm EST / 7pm CET / 10am PT).

It would be great if you could attend this meeting to provide us a short update. 

More information on https://github.com/dipy/dipy/discussions/2398

Thank you!
